cabal-install/main/Main.hs

```diff
@@ -177,8 +178,22 @@ main = do
   -- when writing to stderr and stdout.
   relaxEncodingErrors stdout
   relaxEncodingErrors stderr
-  (args0, args1) <- break (== "--") <$> getArgs
-  mainWorker =<< (++ args1) <$> expandResponse args0
+
+  (args0, args1) <- fmap (drop 1) . break (== "--") <$> getArgs
+
+  let concatWith = DL.foldl1' (\acc x -> acc <> " " <> x)
+  -- Case for @--@, if no @--target-options@ flag is provided, supply
+  -- everything after @--@ as a value for a new @--target-options@,
+  -- otherwise append it to an already existing @--target-options@.
+  if not (null args1) 
+    && not (null args0)
+    && elem (take 1 args0)
+      (map return ["run", "v2-run", "new-run", "test", "v2-test", "new-test", "bench", "v2-bench", "new-bench"])
```

Comment:
I don't like this, however it seems that the actual build phase calls on this part of the code again (I'm actually not sure, [see here](https://github.com/haskell/cabal/blob/c2897a793a3f076f838555f7a890442e1b9a1b31/cabal-install/src/Distribution/Client/CmdRun.hs#L248)). So if I don't handpick the command, it starts complaining i.e. `unrecognized 'act-as-setup' option '--ghc'`, and fails to build.